### PR TITLE
Powder simulation

### PR DESCRIPTION
**Pull request by Marcraven**
# Powder Diffraction Simulation with PyTorch Acceleration

## Summary

This PR introduces a major overhaul of `xrd_simulator` to support **powder diffraction simulation** with **GPU-accelerated rendering** via PyTorch. The refactor enables realistic simulation of polycrystalline samples across all grain size regimes—from nanocrystals to large single crystals—with physically accurate peak broadening models.

### 🚀 PyTorch Backend
- Complete migration from NumPy to PyTorch for all core computations
- Automatic CPU/GPU device selection via new `xrd_simulator.cuda` module
- Memory-efficient batch processing for millions of grains

### 🔬 Multi-Scale Rendering Methods
New rendering system with three physically-motivated methods:

| Method | Grain Size | Description |
|--------|------------|-------------|
| `nano` | < 0.1³ µm³ | Airy disk patterns with Scherrer broadening for nanocrystals |
| `micro` | 0.1³ - pixel³ µm³ | Fast Gaussian profiles for powder patterns |
| `macro` | > pixel³ µm³ | 3D volume projection showing crystal morphology |
| `auto` | All sizes | Automatic method selection based on grain volume |

### 📊 Physical Corrections
- **Lorentz factor**: Proper intensity corrections for diffraction geometry
- **Polarization factor**: X-ray polarization effects
- **Structure factors**: Crystallographic intensity modulation
- New `scattering_factors.py` module with Scherrer formula implementation

### ⚡ Performance Improvements
- Batch rendering of 1M+ grains in seconds on GPU
- Intelligent memory management with automatic batching
- Removed NumPy dependency from hot paths

## Breaking Changes

- `ScatteringUnit` class removed (functionality merged into `Polycrystal` and `Detector`)
- Rendering method names changed: `centroid` → `micro`, `profiles` → `nano`, `volumes` → `macro`
- Minimum Python version: 3.10
- PyTorch 2.5+ required

## Testing

- ✅ **82/82 tests passing**
- New end-to-end tests for powder diffraction with pyFAI peak finding
- New crystallite size broadening validation tests
- CUDA device configuration tests

## Files Changed

- **100 files** modified (+10,139 / -5,764 lines)
- Core modules refactored: `detector.py`, `polycrystal.py`, `motion.py`, `mesh.py`, `laue.py`
- New modules: `cuda.py`, `scattering_factors.py`
- Documentation rebuilt with Sphinx

## Dependencies Updated

```
torch>=2.5.0
numpy>=1.24.0
scipy>=1.11.0
pandas>=2.0.0
```

## Commits (20)

1. Vectorization merge and initial refactoring
2. PyTorch backend implementation through `_diffract`
3. Torch/NumPy equivalence achieved for full simulation pipeline
4. Multi-frame rendering support
5. Powder diffraction mode with PSF convolution
6. Lorentz, polarization, and structure factor corrections
7. Memory-efficient batch processing
8. Removed NumPy from computations
9. Volume projection for macro grains
10. Gaussian interpolation for micro grains
11. Airy disk patterns for nano grains
12. Auto mode with grain-size-based method selection
13. Vectorized `RigidBodyMotion` for batch operations
14. End-to-end powder diffraction tests
15. CUDA device configuration module
16. Scherrer formula implementation
17. Documentation and docstring updates
18. CI/CD workflow updates for Python 3.13
19. Repository cleanup
20. Sphinx documentation rebuild

---

**Ready for review and merge into `main`.**

### CODE REVIEW

**Comment by AxelHenningsson** [comment]
OK, second round here now. First some points from last round:

4. Yes! We need this returned, it may be optional, that is fine! But at least the intersection point zd, yd, the hkl, and then the phase number such that it is possible to do basic annotation of the simulated diffraction patterns. Does every new render call keep appending to the column list with the strings of zd yd etc? I am a bit confused to what this output is to be used for.

5. Just ran the script you linked. I still see nothing in ```detector.frames``` it is an empty list. I am not talking about the keyword for rendering. I am talking about the class [attribute here. ](https://github.com/FABLE-3DXRD/xrd_simulator/blob/71e42d205484a724bab96bd62d6bf91db7e22fe2/xrd_simulator/detector.py#L72) If I recursively search the code base, then "self.frames" is never used. 

And then the new stuff:

1. Could we extend this doc-part a bit:
        - ``'micro'``: Fast Gaussian profiles, ideal for quick analysis and
          medium-sized grains.
        - ``'nano'``: Physically accurate Airy disk patterns from crystallite
          diffraction. Broader for smaller crystallites, delta-like for large
          crystals. This is the most realistic shape for size-broadened peaks.
        - ``'macro'``: Projects 3D crystal volumes, showing morphology and strain
          effects. Requires volume data in ``peaks_dict``.
        - ``'auto'``: Automatically selects the rendering method based on grain
          size, using ``macro_grain_limit`` and ``micro_grain_limit`` thresholds.
It is not clear exactly what each method will do, I feeel this is a major contribution. It deserves a few sentences per method to describe exactly what is going on :)

2. there seem to be a hidden hard cap on gaussian kernel sizes in the rendering. This should be documented up front, it is confusing that the kernel_threshold has no effect at some point.

3. method="macro" on render when arrays are allocated on gpu fails. This is ok / expected. But we need a gracious not implemented method. "method == macro is not implemeted for cuda device=gpu, consider moving arrays to cpu." 

5. General question: Are there support for setting requires_grad=True on some sample arrays and get autodiffs on a loss on the detector function at this point? Something that has been tried out for a simple scenario?

Things are looking good in general. :+1:  I am ready to do the massive merge after these changes. The follow up step is to make a new pypi release. :)

Cheers
Axel

**Comment by Marcraven** [comment]
Thank you for all your effort Axel! <3<3<3

I think this upgrade is worth publishing! Let's wait for the users to give us feedback!

Let's keep in touch!
